### PR TITLE
`GenesisBuild` split

### PR DESCRIPTION
**Pull request by michalkucharczyk**
This PR removes the storage building part from the `GenesisBuild` trait.

`assimilate_storage` and `build_storage` methods were moved to new `GenesisBuildStorage` trait which is implemented only for pallets `GenesisConfig` (and not `RuntimeGenesisConfig`). `PalletGenesisConfig` marker trait was added to facilitate this.

The rationale behind this is as follows: since paritytech/substrate#14306 `GenesisBuild` is implemented for `RuntimeGenesisConfig`. As the `BuildStorage` is also implemented for `RuntimeGenesisConfig` there was is duality now - `RuntiemGenesisConfig` had two `bulid_storage` implementations.

Trait split is implemented in 13d66fd5d0, renaming in ad3b7df4ba.

Step towards: https://github.com/paritytech/polkadot-sdk/issues/25

cumulus companion: paritytech/cumulus#2720

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2961603

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2961605

File: frame/support/src/traits/hooks.rs [hunk 1]
```diff
@@ -359,15 +359,22 @@ pub trait Hooks<BlockNumber> {
...
 pub trait GenesisBuild<T, I = ()>: Default + sp_runtime::traits::MaybeSerializeDeserialize {
 	/// The build function is called within an externalities allowing storage APIs.
 	/// Thus one can write to storage using regular pallet storages.
 	fn build(&self);
+}
 
+#[cfg(feature = "std")]
+/// A trait to define functions to build the  storage for a genesis config, T and I are placeholder
+/// for pallet trait and pallet instance.
+pub trait GenesisBuildStorage<T, I = ()>: GenesisBuild<T, I> {
```
Comment (thiolliere) [inline frame/support/src/traits/hooks.rs:376]: I don't really understand, you but in another trait these method so you can implement `GenesisBuild` on `RuntimeGenesisConfig` ?

I'm not sure it worth it, why not create a new trait for `RuntimeGenesisConfig` specifically, `GenesisBuild` was meant for pallets and thus the signature is quite specific with `<T, I>`, I think it is better to keep traits and implementation of `RuntimeGenesisConfig` apart, no ?

Comment (thiolliere) [reply frame/support/src/traits/hooks.rs:376]: Sorry for my bad suggestion, I agree what other said in the other PR, maybe this PR improve on the situation with a unique proper GenesisBuild, we could eventually get rid of the generics once with have decl_storage removed (is it already ?) and we can keep the other method for not breaking peoples tests in another trait or another way

**Pull request closed without merging**